// nativescript-charts/components/legend/index.android.js
"use strict";
var LegendHorizontalAlignment;
(function (LegendHorizontalAlignment) {
    LegendHorizontalAlignment[LegendHorizontalAlignment["LEFT"] = com.github.mikephil.charting.components.Legend.LegendHorizontalAlignment.LEFT] = "LEFT";
    LegendHorizontalAlignment[LegendHorizontalAlignment["CENTER"] = com.github.mikephil.charting.components.Legend.LegendHorizontalAlignment.CENTER] = "CENTER";
    LegendHorizontalAlignment[LegendHorizontalAlignment["RIGHT"] = com.github.mikephil.charting.components.Legend.LegendHorizontalAlignment.RIGHT] = "RIGHT";
})(LegendHorizontalAlignment = exports.LegendHorizontalAlignment || (exports.LegendHorizontalAlignment = {}));
var LegendVerticalAlignment;
(function (LegendVerticalAlignment) {
    LegendVerticalAlignment[LegendVerticalAlignment["TOP"] = com.github.mikephil.charting.components.Legend.LegendVerticalAlignment.TOP] = "TOP";
    LegendVerticalAlignment[LegendVerticalAlignment["CENTER"] = com.github.mikephil.charting.components.Legend.LegendVerticalAlignment.CENTER] = "CENTER";
    LegendVerticalAlignment[LegendVerticalAlignment["BOTTOM"] = com.github.mikephil.charting.components.Legend.LegendVerticalAlignment.BOTTOM] = "BOTTOM";
})(LegendVerticalAlignment = exports.LegendVerticalAlignment || (exports.LegendVerticalAlignment = {}));
var LegendForm;
(function (LegendForm) {
    LegendForm[LegendForm["SQUARE"] = com.github.mikephil.charting.components.Legend.LegendForm.SQUARE] = "SQUARE";
    LegendForm[LegendForm["CIRCLE"] = com.github.mikephil.charting.components.Legend.LegendForm.CIRCLE] = "CIRCLE";
    LegendForm[LegendForm["LINE"] = com.github.mikephil.charting.components.Legend.LegendForm.LINE] = "LINE";
})(LegendForm = exports.LegendForm || (exports.LegendForm = {}));
